pkg/crypto/certmanager/manager.go

```diff
@@ -0,0 +1,451 @@
+			time.Now().Add(wait), format.Duration(wait))
+		time.Sleep(wait)
+	}
+}
+
+func (cm *CertificateManager) checkRenew() time.Duration {
+	cm.rwMutex.RLock()
+	cert := cm.certificate
+	cm.rwMutex.RUnlock()
+	// First check if the current certificate needs to be renewed.
+	if cert != nil {
+		if interval := cert.timeUntilRenewal(cm.renewBefore); interval > 0 {
+			return interval
+		}
+	}
+	if cm.storer != nil {
+		// Now try and get a certifcate from the remote store, save it and see
+		// if it needs to be renewed.
+		if cert, err := readCert(cm.storer); err != nil {
+			cm.logger.Println(err)
+		} else if expire := cert.timeUntilRenewal(cm.renewBefore); expire > 0 {
+			cm.rwMutex.Lock()
+			cm.certificate = cert
+			cm.rwMutex.Unlock()
+			return expire
+		}
+	}
+	if err := cm.renew(); err != nil {
+		cm.logger.Println(err)
+		return jitteryHour()
+	}
+	expire := cm.certificate.timeUntilRenewal(cm.renewBefore)
+	if expire < time.Hour {
+		expire = jitteryHour()
+	}
+	return expire
+}
+
+func (cm *CertificateManager) fileLoad() error {
+	if cm.certFilename == "" || cm.keyFilename == "" {
+		return nil
+	}
+	cert, err := loadCertificate(cm.certFilename, cm.keyFilename, cm.logger)
+	if err != nil {
+		return err
+	}
+	cm.rwMutex.Lock()
+	defer cm.rwMutex.Unlock()
+	cm.certificate = cert
+	return nil
+}
+
+func (cm *CertificateManager) fileWrite(cert *Certificate) {
+	if err := cm.fileWriteError(cert); err != nil {
+		cm.logger.Println(err)
+	}
+	cm.logger.Printf("wrote certificate to: %s\n", cm.certFilename)
+}
+
+func (cm *CertificateManager) fileWriteError(cert *Certificate) error {
+	if cm.certFilename == "" || cm.keyFilename == "" {
+		return nil
+	}
+	pid := os.Getpid()
+	certFilename := fmt.Sprintf("%s~%d~", cm.certFilename, pid)
+	keyFilename := fmt.Sprintf("%s~%d~", cm.keyFilename, pid)
+	defer os.Remove(certFilename)
+	defer os.Remove(keyFilename)
+	err := ioutil.WriteFile(certFilename, cert.CertPemBlock, 0644)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(keyFilename, cert.KeyPemBlock, 0600)
+	if err != nil {
+		return err
+	}
+	if err := os.Rename(certFilename, cm.certFilename); err != nil {
+		return err
+	}
+	if err := os.Rename(keyFilename, cm.keyFilename); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cm *CertificateManager) getCertificate(hello *tls.ClientHelloInfo) (
+	*tls.Certificate, error) {
+	cm.rwMutex.RLock()
+	defer cm.rwMutex.RUnlock()
+	if cm.certificate == nil {
+		return nil, errors.New("no certificate available")
+	}
+	return &cm.certificate.tlsCert, nil
+}
+
+func (cm *CertificateManager) makeAcmeClient(ctx context.Context) error {
+	if cm.acmeClient != nil {
+		return nil
+	}
+	key, err := makeKey()
+	if err != nil {
+		return err
+	}
+	acmeClient := &acme.Client{
+		Key:          key,
+		DirectoryURL: cm.caDirectoryURL,
+		UserAgent: filepath.Base(os.Args[0]) +
+			"using github.com/Cloud-Foundations/golib/pkg/crypto/certmanager",
```

Comment:
dont like this.. is both undefined and too much info... but I dont have suggestions fromt the top of my head

Reply:
If ever a CA is experiencing issues from this code, their logs will show where to find the code.